### PR TITLE
Add `perf_event_paranoid` dependent capability support

### PR DESCRIPTION
**Pull request by cdelzotti**
## Initial Checklist

- [x] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [x] If part of an EPIC, PR git log contains EPIC number.
- [x] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

Capabilities required by Tracee depend on the `perf_event_paranoid` kernel value.
Before this commit, Tracee required the `CAP_SYS_ADMIN` capability to be certain
to run on every configuration. This commit checks the `perf_event_paranoid` to ensure that
Tracee is running with required capabilities only, therefore reducing the exposition to security
issues.

Fixes: #1815

## Type of change

- [x] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [ ] Code refactor (code improvement and/or code removal)
- [ ] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [inline cmd/tracee-ebpf/capabilities.go:85]: too descriptive.

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [inline cmd/tracee-ebpf/capabilities.go:87]: ditto

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [inline cmd/tracee-ebpf/capabilities.go:89]: ditto

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [inline cmd/tracee-ebpf/capabilities.go:100]: Hum, thats a nice catch =). I would add comment from `kernel/events/core.c` as a comment here (or in that function):

```
/*
 * perf event paranoia level:
 *  -1 - not paranoid at all
 *   0 - disallow raw tracepoint access for unpriv
 *   1 - disallow cpu events for unpriv
 *   2 - disallow kernel profiling for unpriv
 *   4 - disallow all unpriv perf event use
 */
```

or something similar.

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [inline cmd/tracee-ebpf/capabilities.go:104]: We can't write messages to stdout in non fatal errors I'm afraid. We should write to stderr OR, even better, to stdout if debug flag is given.

**Comment by rafaeldtinoco** [comment]
@AlonZivony mind taking a look ?

**Comment by rafaeldtinoco** [comment]
The errors are very likely related to the stdout message (FYIO) @cdelzotti

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [inline cmd/tracee-ebpf/capabilities.go:94]:
```suggestion
	value = strings.TrimSuffix(value, '\n')
```

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [inline cmd/tracee-ebpf/capabilities.go:97]: I think it would be better to use consts for these values,
it will make it self-explanatory.

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [inline cmd/tracee-ebpf/capabilities.go:109]: This one is weird.
You wrote in the issue that you saw that values up to 3 enable the use of `CAP_BPF` and `CAP_PERFMON`.
Why did you change it to be 1?
In my system I have the value of 2, though it works for me with  `CAP_BPF` and `CAP_PERFMON`. Is this a bug?

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [reply cmd/tracee-ebpf/capabilities.go]: I think it should probably be printed if in debug mode.
I hope that in the near future we will have a logger to handle this logic, but you can just change it so you get the debug flag as a argument to this function.

**Review by AlonZivony — CHANGES_REQUESTED** [review]
The main issue I see here is with the maximum level of the paranoid which enables the new capabilities.
I think @rafaeldtinoco is right with his assumption that the prints make the tests fail.

File: cmd/tracee-ebpf/capabilities.go
Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go]: I took the liberty to add the config to the function parameters to check the Debug flag, let me know if I shouldn't.

Comment (AlonZivony) [reply cmd/tracee-ebpf/capabilities.go]: Hmm I think we should only pass the flag for now, to make it less dependent and more testable.

File: cmd/tracee-ebpf/capabilities.go
Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go:109]: Just wanted to be safe as the documentation mentions  `2 - disallow kernel profiling for unpriv`. But it does work qith values up to 3 on my machine.

**Comment by AlonZivony** [comment]
Sorry for the delayed CR, it made you do some work twice :(

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [reply cmd/tracee-ebpf/capabilities.go:109]: I think we can have the assumption that tracee run as a privileged process and user.

**Comment by cdelzotti** [comment]
Okay, I think I took everything into account.

**Comment by rafaeldtinoco** [comment]
@AlonZivony I'll merge this once you think is good, just let me know. Thanks!

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [inline cmd/tracee-ebpf/capabilities.go:107]: You miss here the check of the kernel version, which should be at least 5.8 to use `CAP_BPF` and `CAP_PERFMON` (they don't exist before this version).

Should probably look the following:
```suggestion
        const bpfCapabilitiesMinKernelVersion = "5.8"
        if OSInfo.CompareOSBaseKernelRelease(bpfCapabilitiesMinKernelVersion) <= 0 {
	    if kernelParanoidValue > 3 {
```

**Review by AlonZivony — CHANGES_REQUESTED** [review]
Just noticed the missing check of kernel versions (because of the failed tests).
Add it and it should be good.
Notice you will need to change the tests to check all these scenarios.

File: cmd/tracee-ebpf/capabilities.go
Comment (AlonZivony) [reply cmd/tracee-ebpf/capabilities.go:107]: I will add that when I implemented it, I checked when using `CAP_BPF` and `CAP_PERFMON` if they are available.
If they aren't, I reverted to using `CAP_SYS_ADMIN`.
I think we try at all cost to make tracee work when used, so if there is a way to run it with even higher capabilities then we will use it.
You can see my implementation in commit 6b0cad48.

**Review by AlonZivony — APPROVED** [review]
LGTM

File: cmd/tracee-ebpf/capabilities.go
Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go]: Thank you, how can I modify the tests ?

File: docs/installing/prerequisites.md
Comment (yanivagman) [inline docs/installing/prerequisites.md:31]: Above it seems that you check if paranoid level is more than 3, not 1 like written here. Which one is true?

Comment (cdelzotti) [reply docs/installing/prerequisites.md:31]: 1, forgot to update the documentation, sorry

File: cmd/tracee-ebpf/capabilities.go
Comment (yanivagman) [inline cmd/tracee-ebpf/capabilities.go:126]: So I wonder if this is actually a bug (in Ubuntu?) since the kernel documentation says that CAP_PERFMON should be enough for every operation used with perf: https://www.kernel.org/doc/html/latest/admin-guide/perf-security.html#perf-events-access-control 
Actually, the kernel sources don't even mention levels 3 or 4 as can be seen in the comment here:
https://elixir.bootlin.com/linux/v5.19/source/kernel/events/core.c#L416
So if I understand it right, it's probably an addition of Ubuntu, which is not part of mainline kernel, and they "forgot" to allow the operation with CAP_PERFMON enabled, am I right?

**Review by yanivagman — COMMENTED** [review]
Thanks @cdelzotti 
A few comments/questions and I think we're set

File: cmd/tracee-ebpf/capabilities.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/capabilities.go:126]: They extracted from grsecurity patches:

image

Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/capabilities.go:126]: I think we should assume that there is a deviation in between upstream acceptance of >=2 and grscurity/ubuntu (and possibly other distros ?) having > 2 values.

Comment (yanivagman) [reply cmd/tracee-ebpf/capabilities.go:126]: > They extracted from grsecurity patches:
> 
> image

From when is this commit dated? Maybe they required CAP_SYS_ADMIN before CAP_PERFMON was available (5.8), and simply forgot to add the option for this capability sinve these kernels

File: cmd/tracee-ebpf/capabilities.go
Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go:92]: Hmm, why not, this would allow the caller to print a more precise debug message I guess

File: cmd/tracee-ebpf/capabilities.go
Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go:126]: I think you might be right, as in this discussion (https://patchwork.kernel.org/project/linux-hardening/patch/1469630746-32279-1-git-send-email-jeffv@google.com/) they seem to wish for a `CAP_PERF` that would handle this kind of things.

Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go:126]: > I think we should assume that there is a deviation in between upstream acceptance of >=2 and grscurity/ubuntu (and possibly other distros ?) having > 2 values.

That really looks like it as I noticed that Tracee runs fine on a recompiled kernel with a paranoid value > 3 without `CAP_SYS_ADMIN`

Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/capabilities.go:132]: Yep and I can talk to their kernel team and ask them to address that, and will, but we will have to work in many already releases broken kernels even if they fix that in latest releases I'm afraid.

Comment (cdelzotti) [reply cmd/tracee-ebpf/capabilities.go:132]: Yeah, so I don't really think it changes something to the way this PR addresses the `perf_event_paranoid` value problem

**Review by yanivagman — APPROVED** [review]
I'll let @rafaeldtinoco have another look before merging this, but LGTM, thanks @cdelzotti

**Review by rafaeldtinoco — COMMENTED** [review]
Thanks @yanivagman,

@cdelzotti how do you feel about making the change proposed at:
https://gist.github.com/rafaeldtinoco/9d913fb93a03dd8697b9b678f199baa4

Reasons:

1. We dont have to follow `caps` value (return already says what we are returning each time.
2. Simplifies 2 "if/elses" (no need for else if we already return before.
3. Debug messages are usually prefixed with the session we're currently debugging for and are easier to follow (simpler to know what is the file we tried to read).

feel free to include this change if you agree, or parts of it, and we can merge this. could you also make sure I did not do anything wrong ? I haven't tested the change.

**Comment by rafaeldtinoco** [comment]
Thanks @cdelzotti for your contribution!

**Pull request merged**
